app/controllers/course/learning_map_controller.rb

```diff
@@ -1,15 +1,214 @@
 # frozen_string_literal: true
 class Course::LearningMapController < Course::ComponentController
+  NODE_ID_DELIMITER = '-'
+  NEGATIVE_INF = -1_000_000_000
+
   add_breadcrumb :index, :course_learning_map_path
 
   def index
+    init_conditionals
+
+    respond_to do |format|
+      format.html
+      format.json do
+        render json: { nodes: map_conditionals_to_nodes, can_modify: current_course_user&.staff? }
+      end
+    end
   end
 
-  private
+  def add_parent_node
+    condition = create_condition(parent_and_node_id_pair_params[:parent_node_id])
+    conditional = get_conditional(parent_and_node_id_pair_params[:node_id])
+    condition.conditional = conditional
+
+    if condition.save
+      index
+    else
+      error_response(condition.errors.full_messages)
+    end
+  end
+
+  def remove_parent_node
+    condition = get_condition(parent_and_node_id_pair_params[:parent_node_id],
+                              parent_and_node_id_pair_params[:node_id])
+
+    if condition.destroy
+      index
+    else
+      error_response(condition.errors.full_messages)
+    end
+  end
+
+  def toggle_satisfiability_type
+    conditional = get_conditional(node_params[:node_id])
+
+    if conditional.satisfiability_type.to_s == :all_conditions.to_s
+      conditional.set_at_least_one_condition_satisfiability_type!
+    else
+      conditional.set_all_conditions_satisfiability_type!
+    end
+
+    if conditional.save
+      index
+    else
+      error_response(conditional.errors.full_messages)
+    end
+  end
 
   # @return [Course::LearningMapComponent]
   # @return [nil] If component is disabled.
   def component
     current_component_host[:course_learning_map_component]
   end
+
+  private
+
+  def error_response(errors)
+    respond_to do |format|
+      format.html
+      format.json do
+        render json: { errors: errors }, status: :bad_request
+      end
+    end
+  end
+
+  def init_conditionals
+    @conditionals = Course::Condition.conditionals_for(current_course)
+  end
+
+  def map_conditionals_to_nodes
+    all_node_relations = generate_all_node_relations
+    nodes = generate_nodes_from_conditionals(all_node_relations)
+    generate_node_depths(nodes)
+  end
+
+  def generate_all_node_relations
+    relations = init_all_node_relations
+    node_ids_to_children = relations[:node_ids_to_children]
+    node_ids_to_parents = relations[:node_ids_to_parents]
+
+    @conditionals.each do |conditional|
+      node_id = get_node_id(conditional)
+
+      conditional.conditions.each do |condition|
```

Comment:
Should preload this as well, if necessary.

Reply:
Done